### PR TITLE
[READY] Global Ticket System / Ticket Handling

### PR DESCRIPTION
**Pull request by ZephyrTFA**
This PR adds two major things: a Global Ticket system and a Ticket Handling system.

## Global Ticket System ##
The Global Ticket System, which will now be referred to as either **GTS** or the **DispatchSS**, is the back end of a rather large and semi-complex system in which all jobs are catalogued on initialization of the **GTS** and assigned a role cache and a holder cache.
The role cache stores a list of every job that was initialized and has values assigned to it that dictate what exactly that job can access in regards to the **GTS**. The holder cache stores a list of every mob whose job has a role in the cache, henceforth known as a **Holder**.

## THS / Non-Holder ##
The Ticket Handling System is the front end for the **GTS**. Is provides the interactions to create, modify, and handle tickets for both **Holders** and Non-**Holders**. A Non-**Holder** has the ability to create and submit tickets to identify and report on issues that range from "Sometime Soon" to "Help NOW".

These tickets contain a important information: 
- Creator
- Location
- Title
- Description
- Optional Suspect Information
- Optional Image

Creator and Location are set automatically upon ticket submission, however the Title and Description may be edited freely prior to submission. You may also choose from a range of templates that will automatically fill out the Title, Description, Ticket Type, and Ticket Priority for you; for the panicking assistant on the go.

## THS / Holder ##
A **Holder** has the ability to sign into the **Ticket Browser** which gives them the ability to clock in/out of ticket duty, go active/inactive for their preferred role, and also view all tickets and modify them. A **Holder** has the ability to select a ticket from a list, which is sorted by priority so that the most important tickets are at the top, and view information about it. Should they decide to they can 'Handle' the ticket which will assign them as a handler for the ticket and send a message to the ticket creator that 'X' is handling their ticket.
The ticket handler can choose to 'Resolve' or 'Reject' tickets, which will each send a message to the ticket creator. They can also choose to 'Clear' a ticket, which will reset the ticket back to it's initial state and clear both the Handler data and also the Status of the ticket.

## Antaggery ##
Presently the ONLY thing antags can do is emag their headset to be able to spoof Creator and Location data.

## Dispatch and Control ##
The Dispatch and Control Console, henceforth known as the DCC, is used to manage Tickets and Holders.

The DCC:
- Can see all **Holders** and their currently active ticket.
- Can see all Tickets
- Forcibly reassign tickets
- Change Ticket Priority
- Change Ticket Status

## Images ##
t1
t2
t3
t4
t5
t6
t7
t8
t9
t99
image
image2
https://user-images.githubusercontent.com/12817816/122692011-b006ca00-d200-11eb-97af-6b9c4832cde7.mp4

### CODE REVIEW

**Comment by FlamingLily** [comment]
Holy shit _yes please_

**Comment by YakumoChen** [comment]
I legitimately can't tell if this is supposed to be a system for ahelp tickets, actual in-game work ala IT Department, or some weird mix of both.

**Comment by YakumoChen** [comment]
If it's the latter I am hopeful to see the rise of in-game workplace karens who submit high priority tickets to engineering because someone crowbarred a single floor tile

**Comment by ORCACommander** [comment]
is this intended to replace the old requests console?

**Comment by ZephyrTFA** [comment]
@Gandalf2k15 TM this please :)

File: modular_skyrat/modules/dispatching/dispatch_controller.dm [hunk 1]
```diff
@@ -0,0 +1,516 @@
...
+		if((ui_data_by_mob[mob] && ui_data_by_mob[mob]["mdata"]["holderActive"]) || ignore_active)
+			message_holder(mob, message)
+
+///Takes a user, and a list of the ticket data, and constructs a ticket to then alert type holders and slots it into the master list
+/datum/controller/subsystem/dispatch/proc/ticket_create(mob/user, list/tdata)
+	holder_update()
+	var/datum/dispatch_ticket/ticket = new(user, tdata)
+	user.visible_message("[user] makes a loud chime!")
+	for(var/mob/mob in viewers(world.view, user))
+		SEND_SOUND(mob, '/sound/machines/terminal_success.ogg')
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/dispatch_controller.dm]:
```suggestion
		SEND_SOUND(mob, 'sound/machines/terminal_success.ogg')
```

File: modular_skyrat/modules/dispatching/ticket_queue_action.dm [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+	if(!ishuman(usr)) return
+	SSdispatch.ui_interact(usr, null, "ticket-new")
+
+/obj/item/radio/headset/RightClick(mob/user)
+	var/list/holder_roles = SSdispatch.get_holder_roles(user)
+	if(!holder_roles.len)
+		to_chat(user, span_warning("You are not authorized to access the Dispatch Browser."))
+		return
+	SSdispatch.ui_interact(user, null, "ticket-manage")
+	return COMPONENT_CANCEL_CLICK_RIGHT
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/ticket_queue_action.dm]:
```suggestion
	return COMPONENT_SECONDARY_CANCEL_ATTACK_CHAIN
```

File: modular_skyrat/modules/dispatching/ticket_queue_action.dm [hunk 3]
```diff
@@ -0,0 +1,23 @@
+/datum/action/item_action/dispatch_ticket_new
+	name = "Report Incident/Issue"
+	button_icon_state = "vote"
+
+/datum/action/item_action/dispatch_ticket_new/Trigger()
+	if(!ishuman(usr)) return
+	SSdispatch.ui_interact(usr, null, "ticket-new")
+
+/obj/item/radio/headset/RightClick(mob/user)
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/ticket_queue_action.dm]:
```suggestion
/obj/item/radio/headset/pre_attack_secondary(atom/target, mob/user, params)
```

File: modular_skyrat/modules/dispatching/dispatch_and_control.dm [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+/datum/action/item_action/dispatch_management/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+
+	var/mob/living/carbon/human/user = locate(params["self_ref"])
+	var/mob/living/carbon/human/holder = params["holder_ref"] ? locate(params["holder_ref"]) : null
+	if(!istype(user))
+		CRASH("invalid self_ref. must be a human reference")
```
Comment (tf-4) [inline modular_skyrat/modules/dispatching/dispatch_and_control.dm]:
```suggestion
		CRASH("invalid self_ref. Must be a human reference")
```

File: modular_skyrat/modules/dispatching/dispatch_controller.dm [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+/datum/controller/subsystem/dispatch/proc/message_type_holders(ticket_type, message, ignore_active=FALSE)
+	for(var/mob/mob in job_type_holders[ticket_type])
+		if((ui_data_by_mob[mob] && ui_data_by_mob[mob]["mdata"]["holderActive"]) || ignore_active)
+			message_holder(mob, message)
+
+///Takes a user, and a list of the ticket data, and constructs a ticket to then alert type holders and slots it into the master list
+/datum/controller/subsystem/dispatch/proc/ticket_create(mob/user, list/tdata)
+	holder_update()
+	var/datum/dispatch_ticket/ticket = new(user, tdata)
+	user.visible_message("[user] makes a loud chime!")
```
Comment (tf-4) [inline modular_skyrat/modules/dispatching/dispatch_controller.dm]: I feel like this should be more obvious what it is, people won't know what that means. Maybe something like "[user] makes a loud chime - ticket created successfully!"

```suggestion
	user.visible_message("[user] makes a loud chime - ticket created successfully!")
```

File: modular_skyrat/modules/dispatching/dispatch_datum.dm [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+/datum/dispatch_ticket/proc/handle(mob/user, alert = TRUE)
+	status = SSDISPATCH_TICKET_STATUS_ACTIVE
+	if(handler)
+		handler_past += handler
+	handler = user
+	if(handler && alert)
+		message_creator("Your ticket is being handled by [handler]")
+
+/datum/dispatch_ticket/proc/message_creator(message)
+	to_chat(origin, "<span class='boldnotice'>Ticket Message: [message]</span>")
```
Comment (tf-4) [inline modular_skyrat/modules/dispatching/dispatch_datum.dm]:
```suggestion
	to_chat(origin, span_boldnotice("Ticket Message: [message]"))
```

File: modular_skyrat/modules/dispatching/dispatch_controller.dm (hunk 5 above)
Comment (ZephyrTFA) [reply modular_skyrat/modules/dispatching/dispatch_controller.dm]: Sure, suggest it so I can web commit it?

File: modular_skyrat/modules/dispatching/dispatch_and_control.dm [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+	button_icon_state = "round_end"
+	var/dispatch_type
+	var/current_filter = "Nothing"
+
+/obj/item/radio/headset/Initialize()
+	. = ..()
+	actions += new /datum/action/item_action/dispatch_management(src)
+
+/// Yes I know this is ugly. I dont care. ~ZephyrTFA
+/datum/action/item_action/dispatch_management/Grant(mob/M)
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/dispatch_and_control.dm]:
```suggestion
/datum/action/item_action/dispatch_management/Grant(mob/action_owner)
```

File: modular_skyrat/modules/dispatching/dispatch_and_control.dm [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+	if(!ishuman(M))
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/dispatch_and_control.dm]:
```suggestion
	if(!ishuman(action_owner))
```

File: modular_skyrat/modules/dispatching/dispatch_and_control.dm [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+		return
+	var/mob/living/carbon/human/user = M
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/dispatch_and_control.dm]:
```suggestion
	var/mob/living/carbon/human/user = action_owner
```

File: modular_skyrat/modules/dispatching/dispatch_controller.dm (hunk 5 above)
Comment (tf-4) [reply modular_skyrat/modules/dispatching/dispatch_controller.dm]: Sure.

**Comment by CameronWoof** [comment]
Is any text/alert/sound given when someone creates or sends a ticket? Could someone make a ticket saying "That griefing traitor george melons has me cornered in maint" with a picture of them attached without that griefing traitor george melons knowing it's happening?

**Comment by ZephyrTFA** [comment]
> Is any text/alert/sound given when someone creates or sends a ticket? Could someone make a ticket saying "That griefing traitor george melons has me cornered in maint" with a picture of them attached without that griefing traitor george melons knowing it's happening?

You make a loud chime and anyone who can see you gets a chat message that you submitted a ticket

**Comment by IgiariValkyr** [comment]
Security Sergeants are unable to access dispatch

File: modular_skyrat/modules/dispatching/ticket_queue_action.dm [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	var/list/holder_roles = SSdispatch.get_holder_roles(user)
+	if(!holder_roles.len)
+		to_chat(user, span_warning("You are not authorized to access the Dispatch Browser."))
+		return
+	ui_cache = SSdispatch.ui_interact(user, ui_cache)
+
+/obj/item/radio/headset/Click(location, control, params)
+	. = ..()
+	var/mob/living/carbon/user = loc
+	if(!istype(loc) || loc.ears != src)
```
Comment (ZephyrTFA) [inline modular_skyrat/modules/dispatching/ticket_queue_action.dm]:
```suggestion
	if(!istype(loc) || user.ears != src)
```

**Comment by Gandalf2k15** [comment]
You're gonna hate what I'm about to suggest, but, this has been bothering me for a while. The additional action in my action bar is enough to put me off using this. 

My suggestion is to move this from headsets and onto a modular computer app. Or a physical terminal. As currently, every headset having this component is not practical and negatively affects performance. 

And given action bar real estate is incredibly valuable I think if you can't do the above, add a UI button to headsets instead of an action button.

**Comment by hugmeorelse** [comment]
I've asked for it to be an action button. You need to convince me, not him. @Gandalf2k15 

We want to encourage people to use this once we get a proper guide up and a proper understanding of how it's supposed to be used, once it's ready. If it's not readily visible and available to be used, people will not use it. Action button real estate is precious, but I consider this to be rather important, considering the lack of coordination in a department has proven to be problematic. This will not only be effective for security but other departments like engineering.

If you have suggestions that fit within the parameters and meets the goal I stated, I am happy to hear it out.

Thanks.

**Comment by CameronWoof** [comment]
I think, conceptually, this system would work better as a PDA/tablet program than a sort of... inexplicable use of the headset. It's unclear if you're speaking into the headset, or typing on some kind of... headset-mounted keyboard? Not sure. 

As for action button real estate, this is at least as important as a PDA flashlight, so I think that part's fine. Though I don't like that it seems to take up *two* action buttons.

**Comment by ZephyrTFA** [comment]
This is being closed/drafted while I fix my technical debt

**Comment by ZephyrTFA** [comment]
While I resolve serious technical debt, likely by redesigning this from the ground up, this will be closed. It is still actively being worked on however

**Pull request closed without merging**